boot/bootutil/src/loader.c

```diff
@@ -908,9 +1024,55 @@ boot_copy_image(struct boot_status *bs)
     int first_sector_idx;
     int last_sector_idx;
     int swap_idx;
+    struct image_header *hdr;
+    uint32_t size;
+    uint32_t copy_size;
+    struct image_header tmp_hdr;
+    int rc;
+
+    /* FIXME: just do this if asked by user? */
+
+    size = copy_size = 0;
+
+    hdr = &boot_data.imgs[0].hdr;
+    if (hdr->ih_magic == IMAGE_MAGIC) {
+        copy_size = hdr->ih_hdr_size + hdr->ih_img_size + hdr->ih_tlv_size;
+    }
+
+    hdr = &boot_data.imgs[1].hdr;
+    if (hdr->ih_magic == IMAGE_MAGIC) {
+        size = hdr->ih_hdr_size + hdr->ih_img_size + hdr->ih_tlv_size;
+    }
+
+    if (!size || !copy_size || size == copy_size) {
```

Comment:
> What is || size == copy_size for? Is it because we might have the same image's header copied into both slots at the same time? What if the two images are the same size?

Yeah, I think this is tricky. The comparison really tests if it *could be* the same image (same size). If that's the case the scratch is checked and if there is a valid header there, its size if used (assuming both slots' headers where the same). 

> Is it guaranteed that the scratch bytes at the ih_magic offset will be something different than IMAGE_MAGIC when this path is reached?

The only chance of having invalid data or same data in the slot's headers is when doing the swap operation of the sector that has the header. Since scratch is the first thing erased on a swap I assume that if it has the `IMAGE_MAGIC`, the header swap was in progress. If it doesn't have the `IMAGE_MAGIC` it's ignored and everything proceeds as if both images were actually the same size. I have to give more thought on what could go wrong here.